voip_utils/sip.py

```diff
@@ -283,3 +329,280 @@ def _parse_sip(
+                        f"ACK {self._request_uri} SIP/2.0",
+                        f"Via: SIP/2.0/UDP {self._source_endpoint.host}:{self._source_endpoint.port}",
+                        f"From: <sip:{self._source_endpoint.username}@{self._source_endpoint.host}:{self._source_endpoint.port}>",
+                        f"To: {to_header}",
+                        f"Call-ID: {self._call_id}",
+                        "CSeq: 50 ACK",
+                        "User-Agent: test-agent 1.0",
+                        "Content-Length: 0",
+                    ]
+                    ack_text = _CRLF.join(ack_lines) + _CRLF
+                    ack_bytes = ack_text.encode("utf-8")
+                    self.transport.sendto(
+                        ack_bytes, (self._dest_endpoint.host, self._dest_endpoint.port)
+                    )
+
+                    # The call been answered, proceed with desired action here
+                    self.on_call(
+                        CallInfo(
+                            caller_ip=self._dest_endpoint.host,
+                            caller_uri=self._dest_endpoint.sip_uri,
+                            caller_name=self._dest_endpoint.description,
+                            caller_sip_port=self._dest_endpoint.port,
+                            caller_rtp_port=remote_rtp_port,
+                            server_ip=self._dest_endpoint.host,
+                            headers=headers,
+                            opus_payload_type=opus_payload_type,  # Should probably update this to eventually support more codecs
+                        )
+                    )
+        except Exception:
+            _LOGGER.exception("Unexpected error handling SIP response")
+
+    @abstractmethod
+    def on_call(self, call_info: CallInfo):
+        """Handle outgoing calls."""
+
+    def hang_up(self):
+        """Hang up the call when finished"""
+        bye_lines = [
+            f"BYE {self._request_uri} SIP/2.0",
+            f"Via: SIP/2.0/UDP {self._source_endpoint.host}:{self._source_endpoint.port}",
+            f"From: <sip:{self._source_endpoint.username}@{self._source_endpoint.host}:{self._source_endpoint.port}>",
+            f"To: <sip:{self._dest_endpoint.username}@{self._dest_endpoint.host}:{self._dest_endpoint.port}>"
+            f"Call-ID: {self._call_id}",
+            "CSeq: 51 BYE",
+            "User-Agent: test-agent 1.0",
+            "Content-Length: 0",
+            "",
+        ]
+        _LOGGER.debug("Hanging up...")
+        bye_text = _CRLF.join(bye_lines) + _CRLF
+        bye_bytes = bye_text.encode("utf-8")
+        self.transport.sendto(
+            bye_bytes, (self._dest_endpoint.host, self._dest_endpoint.port)
+        )
+
+        self.transport.close()
+        self.transport = None
+
+    def connection_lost(self, exc):
+        """Signal wait_closed when transport is completely closed."""
+        self._loop.call_soon_threadsafe(self._closed_event.set)
```

Comment:
Isn't connection_lost called from within the event loop?
```suggestion
        self._closed_event.set()
```